setup/setup_p2.sh

```diff
@@ -1,39 +1,117 @@
+if [ ! -f ~/.ssh/aws-key-$name.pem ] 
 then
-	aws ec2 create-key-pair --key-name aws-key --query 'KeyMaterial' --output text > ~/.ssh/aws-key.pem
-	chmod 400 ~/.ssh/aws-key.pem
+	aws ec2 create-key-pair --key-name aws-key-$name --query 'KeyMaterial' --output text > ~/.ssh/aws-key-$name.pem
+	chmod 400 ~/.ssh/aws-key-$name.pem
 fi
 
-export instanceId=`aws ec2 run-instances --image-id ami-bc508adc --count 1 --instance-type p2.xlarge --key-name aws-key --security-group-ids $securityGroupId --subnet-id $subnetId --associate-public-ip-address --block-device-mapping "[ { \"DeviceName\": \"/dev/sda1\", \"Ebs\": { \"VolumeSize\": 128, \"VolumeType\": \"gp2\" } } ]" --query 'Instances[0].InstanceId' --output text`
+export instanceId=`aws ec2 run-instances --image-id $ami --count 1 --instance-type p2.xlarge --key-name aws-key-$name --security-group-ids $securityGroupId --subnet-id $subnetId --associate-public-ip-address --block-device-mapping "[ { \"DeviceName\": \"/dev/sda1\", \"Ebs\": { \"VolumeSize\": 128, \"VolumeType\": \"gp2\" } } ]" --query 'Instances[0].InstanceId' --output text`
+aws ec2 create-tags --resources $instanceId --tags --tags Key=Name,Value=$name-gpu-machine
 export allocAddr=`aws ec2 allocate-address --domain vpc --query 'AllocationId' --output text`
 
 echo Waiting for instance start...
 aws ec2 wait instance-running --instance-ids $instanceId
 sleep 10 # wait for ssh service to start running too
 export assocId=`aws ec2 associate-address --instance-id $instanceId --allocation-id $allocAddr --query 'AssociationId' --output text`
 export instanceUrl=`aws ec2 describe-instances --instance-ids $instanceId --query 'Reservations[0].Instances[0].PublicDnsName' --output text`
-echo securityGroupId=$securityGroupId
-echo subnetId=$subnetId
-echo instanceId=$instanceId
-echo instanceUrl=$instanceUrl
-echo Connect: ssh -i ~/.ssh/aws-key.pem ubuntu@$instanceUrl
+export ebsVolume=`aws ec2 describe-instance-attribute --instance-id $instanceId --attribute  blockDeviceMapping  --query BlockDeviceMappings[0].Ebs.VolumeId --output text`
+
+# reboot instance, because I was getting "Failed to initialize NVML: Driver/library version mismatch" 
+# error when running the nvidia-smi command
+# see also http://forums.fast.ai/t/no-cuda-capable-device-is-detected/168/13
+aws ec2 reboot-instances --instance-ids $instanceId
+
+# save commands to file 
+echo \# Connect to your instance: > $name-commands.txt # overwrite existing file
+echo ssh -i ~/.ssh/aws-key-$name.pem ubuntu@$instanceUrl >> $name-commands.txt
+echo \# Stop your instance: : >> $name-commands.txt 
+echo aws ec2 stop-instances --instance-ids $instanceId  >> $name-commands.txt
+echo \# Start your instance: >> $name-commands.txt 
+echo aws ec2 start-instances --instance-ids $instanceId  >> $name-commands.txt
+echo Reboot your instance: >> $name-commands.txt 
+echo aws ec2 reboot-instances --instance-ids $instanceId  >> $name-commands.txt
+echo "" 
+# export vars to be sure
```

Comment:
Define a version for this setup, as this script is already incompatible with what people have running, and further changes will further make it incompatible.

At what point should this be done in cloudformation?